airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/stream.py

```diff
@@ -0,0 +1,467 @@
+# MIT License
+#
+# Copyright (c) 2020 Airbyte
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+
+from abc import ABC, abstractmethod
+from copy import deepcopy
+from datetime import datetime
+import json
+import concurrent
+from operator import itemgetter
+from typing import Any, Iterable, Iterator, List, Mapping, MutableMapping, Optional, Tuple, Union
+from traceback import format_exc
+from airbyte_cdk.models.airbyte_protocol import SyncMode
+from fnmatch import fnmatch
+from .fileclient import FileClient
+from .filereader import FileReaderCsv
+
+from airbyte_cdk.logger import AirbyteLogger
+from airbyte_cdk.sources.streams import Stream
+
+
+JSON_TYPES = ["string", "number", "integer", "object", "array", "boolean", "null"]
+
+
+class ConfigurationError(Exception):
+    """Client mis-configured"""
+
+
+class FileStream(Stream, ABC):
+
+    format_filereader_map = {
+        'csv': FileReaderCsv,
+        # 'parquet': FileReaderParquet,
+        # etc.
+    }
+    # TODO: make these user configurable in spec.json
+    ab_additional_col = "_ab_additional_properties"
+    ab_last_mod_col = "_ab_source_file_last_modified"
+    ab_file_name_col = "_ab_source_file_url"
+    datetime_format_string = "%Y-%m-%dT%H:%M:%S%z"
+
+    def __init__(self, dataset_name: str, provider: dict, format: dict, path_patterns: List[str], schema: str = None):
+        """
+        :param dataset_name: table name for this stream
+        :type dataset_name: str
```

Comment:
Mostly force of habit from pre-Typing days and for type surfacing in auto-documentation like [Sphinx](https://docs.readthedocs.io/en/stable/intro/getting-started-with-sphinx.html) or [MkDocs](https://docs.readthedocs.io/en/stable/intro/getting-started-with-mkdocs.html).
I'll remove them since there's no value here (and change my automatic docstring generator settings 😆 )